content/blog/2023/2023-11-20-design-patterns.md

```diff
@@ -0,0 +1,641 @@
+
+    companion object {
+        private val instance: MySingleton by lazy { MySingleton() }
+
+        fun getInstance(): MySingleton {
+            return instance
+        }
+    }
+
+    // Singleton properties and methods go here
+
+    fun doSomething() {
+        println("Singleton is doing something")
+    }
+}
+```
+
+To use the singleton:
+
+```kotlin
+val singletonInstance = MySingleton.getInstance()
+singletonInstance.doSomething()
+```
+
+By using `by lazy`, the `instance` is created only when it's first accessed, making it a lazy-initialized singleton.
+
+## Adapter Design Pattern
+
+The Adapter design pattern allows the interface of an existing class to be used as another interface. It is often used to make existing classes work with others without modifying their source code.
+
+In Kotlin, we can implement the Adapter pattern using either class-based or object-based adapters.
+
+Here's an example of the class-based Adapter pattern:
+
+```kotlin
+// Target interface that the client expects
+interface Target {
+    fun request()
+}
+
+// Adaptee (the class to be adapted)
+class Adaptee {
+    fun specificRequest() {
+        println("Specific request")
+    }
+}
+
+// Class-based Adapter
+class ClassAdapter(private val adaptee: Adaptee) : Target {
+    override fun request() {
+        adaptee.specificRequest()
+    }
+}
+
+// Client code
+fun main() {
+    val adaptee = Adaptee()
+    val adapter: Target = ClassAdapter(adaptee)
+
+    adapter.request()
+}
+
+```
+
+In this example, `Target` is the interface expected by the client. `Adaptee` is the class that needs to be adapted. The `ClassAdapter` class is the adapter that implements the `Target` interface and delegates the `request()` method to the `specificRequest()` method of the `Adaptee` class.
+
+Here's an example of the object-based Adapter pattern using composition:
+
+```kotlin
+// Target interface that the client expects
+interface Target {
+    fun request()
+}
+
+// Adaptee (the class to be adapted)
+class Adaptee {
+    fun specificRequest() {
+        println("Specific request")
+    }
+}
+
+// Object-based Adapter
+class ObjectAdapter(private val adaptee: Adaptee) : Target {
+    override fun request() {
+        adaptee.specificRequest()
+    }
+}
+
+// Client code
+fun main() {
+    val adaptee = Adaptee()
+    val adapter: Target = ObjectAdapter(adaptee)
+
+    adapter.request()
+}
+
+```
+
+In this example, the `ObjectAdapter` class contains an instance of the `Adaptee` class and delegates the `request()` method to the `specificRequest()` method of the `Adaptee` class.
+
+## Decorator Design Pattern
+
+The decorator design pattern allows behavior to be added to an individual object, either statically or dynamically without affecting the behavior of other objects from the same class. In Kotlin, we can implement the decorator pattern using interfaces and classes.
+
+Here's a simple example of the decorator pattern in Kotlin:
```

Comment:
The example shows very much how you would do it in Java. Isn't there a way to use extension functions to implement Decorators in Kotlin?